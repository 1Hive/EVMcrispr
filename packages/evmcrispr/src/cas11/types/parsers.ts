import type { Parser } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

import type { Node } from './ast';

export type NodeParser<T = Node> = Parser<T, string, NodeParserState>;
