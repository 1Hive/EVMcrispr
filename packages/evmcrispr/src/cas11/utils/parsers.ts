import type { Err } from 'arcsecond';

import type { NodeParserState } from '../parsers/utils';

export const buildParserError = (
  { index, error }: Err<string, null> | Err<string, NodeParserState>,
  type: string,
  msg?: string,
): string => {
  const splitRes = error.split('got');
  const wrongValueEncountered =
    splitRes.length === 2 ? splitRes[1].trim() : null;

  const errorMessage = msg
    ? `${msg}${wrongValueEncountered ? `, got ${wrongValueEncountered}` : ''}`
    : error.split('): ')[1];
  return `${type}(col: ${index}): ${errorMessage}`;
};

export const getIncorrectReceivedValue = (errorMsg: string): string => {
  const splitRes = errorMsg.split('got ');

  if (splitRes.length === 2) {
    return `, got ${splitRes[1].trim()}`;
  }

  return '';
};
