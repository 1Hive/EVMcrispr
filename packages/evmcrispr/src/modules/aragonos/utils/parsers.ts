import type { PublicClient } from "viem";
import { parseAbiItem } from "viem";

import type { Address } from "../../../types";
import type { ParsedApp, Repo } from "../types";
import { getSystemApp, isSystemApp } from "./interfaces";

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return "";
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : "";
};

const fetchImplementationAddress = (
  appAddress: Address,
  client: PublicClient,
): Promise<Address> => {
  return client.readContract({
    address: appAddress,
    abi: [
      parseAbiItem("function implementation() public view returns (address)"),
    ],
    functionName: "implementation",
  });
};

export const parseApp = async (
  // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
  app: any,
  client: PublicClient,
): Promise<ParsedApp> => {
  const { address, appId, roles, version } = app;
  const { name: repoName, registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress:
      codeAddress ?? (await fetchImplementationAddress(address, client)),
    contentUri,
    name,
    registryName: registry?.name,
    roles: (roles as any[]).map((role) => ({
      ...role,
      roleHash: role["hash"] ?? role["roleHash"],
    })),
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};
