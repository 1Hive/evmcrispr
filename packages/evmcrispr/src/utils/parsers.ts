import { getSystemApp, isSystemApp } from '.';
import type { ParsedApp, Repo } from '../types';

export const parseAppArtifactName = (name: string): string => {
  if (!name) {
    return '';
  }
  // Split by the first '.' occurrence only.
  const parsedName = name.split(/\.(.+)/);

  return parsedName.length > 1 ? parsedName[1] : '';
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseApp = (app: any): ParsedApp => {
  const { address, appId, implementation, repoName, roles, version } = app;
  const { registry } = app.repo || {};
  const { codeAddress, artifact: rawArtifact, contentUri } = version || {};
  let artifact, name;

  if (isSystemApp(appId)) {
    const systemApp = getSystemApp(appId)!;
    artifact = systemApp.artifact;
    name = systemApp.name;
  } else {
    artifact = JSON.parse(rawArtifact ?? null);
    name = repoName;
  }

  return {
    address,
    appId,
    artifact,
    codeAddress: codeAddress ?? implementation.address,
    contentUri,
    name,
    registryName: registry?.name,
    roles,
  };
};

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const parseRepo = (repo: any): Repo => {
  const { artifact: rawArtifact, contentUri, codeAddress } = repo.lastVersion;

  return {
    artifact: JSON.parse(rawArtifact),
    contentUri,
    codeAddress,
  };
};

export const timeUnits: { [key: string]: number } = {
  s: 1,
  m: 60,
  h: 3600,
  d: 86400,
  w: 604800,
  mo: 2592000,
  y: 31536000,
};
