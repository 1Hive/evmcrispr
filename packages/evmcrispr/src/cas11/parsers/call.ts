import type { Parser } from 'arcsecond';
import {
  char,
  choice,
  coroutine,
  letters,
  possibly,
  recursiveParser,
} from 'arcsecond';

import type { StringLiteralNode } from '../types';
import { NodeType } from '../types';

import {
  addressParser,
  probableIdentifierParser,
  variableIdentifierParser,
} from './primaries';
import { argumentsParser } from './expression';
import { helperFunctionParser } from './helper';

export const callExpressionParser: Parser<any, string, any> = recursiveParser(
  () =>
    coroutine(function* () {
      const target = yield choice([
        addressParser,
        variableIdentifierParser,
        helperFunctionParser,
        probableIdentifierParser,
      ]);

      let res = yield char(':');
      let callExpressionNode: any = target;

      do {
        const callee = yield letters.map(
          (value): StringLiteralNode => ({
            type: NodeType.StringLiteral,
            value,
          }),
        );

        const args = yield argumentsParser;

        callExpressionNode = {
          type: NodeType.CallExpression,
          target: callExpressionNode,
          callee,
          args,
        };

        res = yield possibly(char(':'));
      } while (res);

      return callExpressionNode;
    }),
);
