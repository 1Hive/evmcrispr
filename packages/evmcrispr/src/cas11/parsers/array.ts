import type { Parser } from 'arcsecond';
import {
  between,
  char,
  coroutine,
  recursiveParser,
  sequenceOf,
} from 'arcsecond';

import type { ArrayExpressionNode } from '../types';
import { NodeType } from '../types';

import { argumentExpressionParser } from './expression';

import { commaSeparated, optionalWhitespace } from './utils';

export const arrayExpressionParser: Parser<any, string, any> = recursiveParser(
  () =>
    between(sequenceOf([char('['), optionalWhitespace]))(
      sequenceOf([optionalWhitespace, char(']')]),
    )(commaSeparated(argumentExpressionParser)).map(
      (elements: any): ArrayExpressionNode => {
        return {
          type: NodeType.ArrayExpression,
          elements,
        };
      },
    ),
);

export const arrayExpresionParser = recursiveParser(() =>
  coroutine(function* () {
    yield char('[');
    yield optionalWhitespace;
    const elements = yield commaSeparated(argumentExpressionParser);
    yield optionalWhitespace;
    yield char(']');

    return {
      type: NodeType.ArrayExpression,
      elements,
    };
  }),
);
