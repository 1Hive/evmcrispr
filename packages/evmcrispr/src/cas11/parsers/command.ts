import type { Parser } from 'arcsecond';
import {
  choice,
  coroutine,
  endOfInput,
  many1,
  recursiveParser,
  regex,
} from 'arcsecond';

import type { CommandExpressionNode, CommandIdentifierNode } from '../types';
import { NodeType } from '../types';

import { expressionParser } from './expression';
import { endOfLine, optionalWhitespace, whitespace } from './utils';

const commandNameRegex =
  /^(?:(?<module>[a-zA-Z-]{1,63}(?<!-))(?::))?(?<command>[a-zA-Z-]{1,63}(?<!-))/;
const commandNameParser = regex(commandNameRegex).map(
  (value): CommandIdentifierNode => {
    const res = commandNameRegex.exec(value);
    // eslint-disable-next-line @typescript-eslint/no-non-null-asserted-optional-chain
    const { module, command } = res?.groups!;

    const c: CommandIdentifierNode = {
      type: NodeType.CommandIdentifier,
      value: command,
    };

    c.value = command;
    if (module) c.module = module;

    return c;
  },
);

const commandArgsParser = many1(
  coroutine(function* () {
    yield whitespace;

    const args = yield expressionParser;

    return args;
  }),
);

export const endOfCommandParser = choice([endOfLine, endOfInput]);

export const commandExpressionParser: Parser<
  CommandExpressionNode,
  string,
  any
> = recursiveParser(() =>
  coroutine(function* () {
    yield optionalWhitespace;

    const name = yield commandNameParser;

    const args = yield commandArgsParser;

    yield optionalWhitespace;

    yield endOfCommandParser;

    return {
      type: NodeType.CommandExpression,
      name,
      args,
    } as unknown as CommandExpressionNode;
  }),
);
