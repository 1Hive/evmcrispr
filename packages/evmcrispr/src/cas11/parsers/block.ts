import type { Parser } from 'arcsecond';
import { char, coroutine, recursiveParser } from 'arcsecond';

import type { BlockExpressionNode } from '../types';
import { NodeType } from '../types';
import { commandExpressionParser } from './command';
import {
  endOfLine,
  optionalEmptyLines,
  optionalWhitespace,
  surroundedBy,
} from './utils';

export const blockExpressionParser: Parser<BlockExpressionNode, string, any> =
  recursiveParser(() =>
    coroutine(function* () {
      yield char('(');

      yield optionalWhitespace;

      yield endOfLine;

      const scopedCommands = yield optionalEmptyLines(commandExpressionParser);

      yield surroundedBy(optionalWhitespace)(char(')'));

      return {
        type: NodeType.BlockExpression,
        body: scopedCommands,
      } as unknown as BlockExpressionNode;
    }),
  );
