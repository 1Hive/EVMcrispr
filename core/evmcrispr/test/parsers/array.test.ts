import type { Case } from '@1hive/evmcrispr-test-common';
import { runParser, runParserError } from '@1hive/evmcrispr-test-common';

import { arrayExpressionParser } from '../../src/parsers/array';

describe.concurrent('Parsers - array', () => {
  describe.concurrent.each<Case>([
    { title: '', value: '[12, "a string"]' },
    {
      title: 'with spaces at both ends',
      value: '[    1, "a text string",    3    ]',
    },
    {
      title: 'with nested arrays',
      value:
        '[145e18y, @token(DAI), false, ["a string", anIdentifier, [1, 2, [aDeepDeepIdentifier.open]],  $variable], $fDAIx::host()]',
    },
  ])('', ({ title, value }) => {
    it(`should parse an array ${title} correctly`, () => {
      const parsedValue = runParser(arrayExpressionParser, value);

      expect(parsedValue).toMatchSnapshot();
    });
  });

  describe.concurrent.each<Case>([
    {
      title: 'with multiple primary values between commas',
      value: '[1,multiple values between commas, false]',
    },
    { title: 'with empty elements', value: '[12e14w, ,,]' },
    { title: 'without closing brackets', value: '[12e14w, "asdas"' },
  ])('', ({ title, value }) => {
    it(`should fail when parsing an array ${title}`, () => {
      const error = runParserError(arrayExpressionParser, value);

      expect(error).toMatchSnapshot();
    });
  });
});
