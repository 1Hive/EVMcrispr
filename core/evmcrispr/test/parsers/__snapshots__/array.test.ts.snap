// Vitest Snapshot v1

exports[`Parsers - array > should fail when parsing an array with empty elements 1`] = `"ArrayParserError(1:9): Expecting a valid expression, got ','"`;

exports[`Parsers - array > should fail when parsing an array with multiple primary values between commas 1`] = `"ArrayParserError(1:12): Expecting character ']', got 'v'"`;

exports[`Parsers - array > should fail when parsing an array without closing brackets 1`] = `"ArrayParserError(1:16): Expecting character ']', but got end of input."`;

exports[`Parsers - array > should parse an array  correctly 1`] = `
{
  "elements": [
    {
      "loc": {
        "end": {
          "col": 3,
          "line": 1,
        },
        "start": {
          "col": 1,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "12",
    },
    {
      "loc": {
        "end": {
          "col": 15,
          "line": 1,
        },
        "start": {
          "col": 5,
          "line": 1,
        },
      },
      "type": "StringLiteral",
      "value": "a string",
    },
  ],
  "loc": {
    "end": {
      "col": 16,
      "line": 1,
    },
    "start": {
      "col": 0,
      "line": 1,
    },
  },
  "type": "ArrayExpression",
}
`;

exports[`Parsers - array > should parse an array with nested arrays correctly 1`] = `
{
  "elements": [
    {
      "loc": {
        "end": {
          "col": 8,
          "line": 1,
        },
        "start": {
          "col": 1,
          "line": 1,
        },
      },
      "power": 18,
      "timeUnit": "y",
      "type": "NumberLiteral",
      "value": "145",
    },
    {
      "args": [
        {
          "loc": {
            "end": {
              "col": 20,
              "line": 1,
            },
            "start": {
              "col": 17,
              "line": 1,
            },
          },
          "type": "ProbableIdentifier",
          "value": "DAI",
        },
      ],
      "loc": {
        "end": {
          "col": 21,
          "line": 1,
        },
        "start": {
          "col": 10,
          "line": 1,
        },
      },
      "name": "token",
      "type": "HelperFunctionExpression",
    },
    {
      "loc": {
        "end": {
          "col": 28,
          "line": 1,
        },
        "start": {
          "col": 23,
          "line": 1,
        },
      },
      "type": "BoolLiteral",
      "value": false,
    },
    {
      "elements": [
        {
          "loc": {
            "end": {
              "col": 41,
              "line": 1,
            },
            "start": {
              "col": 31,
              "line": 1,
            },
          },
          "type": "StringLiteral",
          "value": "a string",
        },
        {
          "loc": {
            "end": {
              "col": 55,
              "line": 1,
            },
            "start": {
              "col": 43,
              "line": 1,
            },
          },
          "type": "ProbableIdentifier",
          "value": "anIdentifier",
        },
        {
          "elements": [
            {
              "loc": {
                "end": {
                  "col": 59,
                  "line": 1,
                },
                "start": {
                  "col": 58,
                  "line": 1,
                },
              },
              "type": "NumberLiteral",
              "value": "1",
            },
            {
              "loc": {
                "end": {
                  "col": 62,
                  "line": 1,
                },
                "start": {
                  "col": 61,
                  "line": 1,
                },
              },
              "type": "NumberLiteral",
              "value": "2",
            },
            {
              "elements": [
                {
                  "loc": {
                    "end": {
                      "col": 89,
                      "line": 1,
                    },
                    "start": {
                      "col": 65,
                      "line": 1,
                    },
                  },
                  "type": "ProbableIdentifier",
                  "value": "aDeepDeepIdentifier.open",
                },
              ],
              "loc": {
                "end": {
                  "col": 90,
                  "line": 1,
                },
                "start": {
                  "col": 64,
                  "line": 1,
                },
              },
              "type": "ArrayExpression",
            },
          ],
          "loc": {
            "end": {
              "col": 91,
              "line": 1,
            },
            "start": {
              "col": 57,
              "line": 1,
            },
          },
          "type": "ArrayExpression",
        },
        {
          "loc": {
            "end": {
              "col": 103,
              "line": 1,
            },
            "start": {
              "col": 94,
              "line": 1,
            },
          },
          "type": "VariableIdentifier",
          "value": "$variable",
        },
      ],
      "loc": {
        "end": {
          "col": 104,
          "line": 1,
        },
        "start": {
          "col": 30,
          "line": 1,
        },
      },
      "type": "ArrayExpression",
    },
    {
      "args": [],
      "loc": {
        "end": {
          "col": 120,
          "line": 1,
        },
        "start": {
          "col": 106,
          "line": 1,
        },
      },
      "method": "host",
      "target": {
        "loc": {
          "end": {
            "col": 112,
            "line": 1,
          },
          "start": {
            "col": 106,
            "line": 1,
          },
        },
        "type": "VariableIdentifier",
        "value": "$fDAIx",
      },
      "type": "CallExpression",
    },
  ],
  "loc": {
    "end": {
      "col": 121,
      "line": 1,
    },
    "start": {
      "col": 0,
      "line": 1,
    },
  },
  "type": "ArrayExpression",
}
`;

exports[`Parsers - array > should parse an array with spaces at both ends correctly 1`] = `
{
  "elements": [
    {
      "loc": {
        "end": {
          "col": 6,
          "line": 1,
        },
        "start": {
          "col": 5,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "1",
    },
    {
      "loc": {
        "end": {
          "col": 23,
          "line": 1,
        },
        "start": {
          "col": 8,
          "line": 1,
        },
      },
      "type": "StringLiteral",
      "value": "a text string",
    },
    {
      "loc": {
        "end": {
          "col": 29,
          "line": 1,
        },
        "start": {
          "col": 28,
          "line": 1,
        },
      },
      "type": "NumberLiteral",
      "value": "3",
    },
  ],
  "loc": {
    "end": {
      "col": 34,
      "line": 1,
    },
    "start": {
      "col": 0,
      "line": 1,
    },
  },
  "type": "ArrayExpression",
}
`;
